src/main/java/org/ibp/api/java/impl/middleware/dataset/DatasetServiceImpl.java

```diff
@@ -844,4 +863,90 @@ private void addLocationIdVariable(final List<MeasurementVariable> environmentDe
 		locationIdVariable.setName(LOCATION_ID_VARIABLE_NAME);
 		environmentDetailAndConditionVariables.add(0, locationIdVariable);
 	}
+
+	@Override
+	public void replaceObservationUnitsEntry(final int studyId, final int datasetId, final ObservationUnitEntryReplaceRequest request) {
+		this.studyValidator.validate(studyId, true);
+		this.datasetValidator.validateDataset(studyId, datasetId);
+		this.datasetValidator.validatePlotDatasetType(datasetId);
+		final BindingResult errors = new MapBindingResult(new HashMap<String, String>(), Integer.class.getName());
+		if (this.studyService.studyHasGivenDatasetType(studyId, DatasetTypeEnum.MEANS_DATA.getId())) {
```

Comment:
We can implement a `StudyValidator#validateDoesNotHaveMeansDataset` and `StudyValidator#validateHasNoCrossesOrSelections` and use here and reused in `StudyGermplasmServiceImpl#replaceStudyGermplasm` (extract out of `StudyGermplasmValidator`)